Ensure serial output code match the intention of the developer

X-Ryl669 opened the pull request:
### Description

To follow the current serial code rewrite, here's a PR to fix some of the serial macro's calling error (like using `SERIAL_ECHO` where `SERIAL_CHAR` was intended, ...).
I've documented the **SERIAL_XXX** macros so it's clear which one to use now.
Now, the expectation is that `SERIAL_ECHO` actually **echo** the argument passed in by converting to a human readable string. So `uint8_t a = 32; SERIAL_ECHO(a);` will output `32` and not ` `.
`SERIAL_CHAR` must be used to output a character as-is, thus `SERIAL_CHAR(a);` will output ` `.

With this change, it's no more required to cast a parameter for `SERIAL_ECHO` if it's a `uint8` or a `int8`, so it's one less trouble/pain to remember for developers. 

I've changed the default base for serial's print method to use decimal everywhere. Ideally, the `base == 0` case will disappear, since it was an hack. I'll probably do that later on.

I'll also remove the variadic SERIAL macro to replace them by variadic templates. The main advantage I see is actually reducing the build size since a single implementation of the template will be in the firmware, instead of a multiple time the bunch of instructions the macros are generating. It'll also helps the code analyzers (like Intellisense / clangserver) since they fails to develop the macros correctly.

It also includes a fix that I've missed in my previous PR for those using a MMU serial on AVR. 

### Requirements

None
### Benefits

See above
### Configurations

All
### Related Issues

#20783 (last comment at time of writing)

X-Ryl669 committed:
Ensure serial output code match the intention of the developer

X-Ryl669 committed:
Fix building issues

X-Ryl669 committed:
Prevent compilation of code using zero base for now
Can easily be tested by adding "SERIAL_PRINT((uint8_t)3, 0)" somewhere in a CPP file, it'll fail to build

X-Ryl669 committed:
Reduce serial impact and make it harder to compile bad code

X-Ryl669 commented:
Ok, continuing the work here. I've removed base == 0 case, and make it so, if found in the code, the compiler will break instead of compiling buggy code. 

I've also removed many (but not all) SERIAL macro and replaced them by template instead, leading to:
1. Huge code size saving
2. No more limitation in the number of arguments for the macros
3. Compile-time checking of parameters (well sort-of, right now, only ECHOPAIR would benefit from this, but I've not enabled for ECHOPAIR since it make code larger... grrr...)
Before:
```
RAM:   [=======   ]  74.6% (used 48880 bytes from 65536 bytes)
Flash: [====      ]  44.7% (used 234540 bytes from 524288 bytes)
```
After:
```
RAM:   [=======   ]  74.6% (used 48880 bytes from 65536 bytes)
Flash: [====      ]  44.5% (used 233380 bytes from 524288 bytes)
```

Need to sort out the combination problem of SERIAL_ECHOPAIR... (and fix the building issues)

GMagician commented on Marlin/src/gcode/calibrate/M48.cpp:
```diff
@@ -240,7 +240,7 @@ void GcodeSuite::M48() {
       if (verbose_level > 1) {
-        SERIAL_ECHO((int)(n + 1));
+        SERIAL_ECHO(n + 1);
         SERIAL_ECHOPAIR(" of ", (int)n_samples);
```
is int cast really needed anymore?

X-Ryl669 replied:
I'm not sure. Currently, SERIAL_ECHOPAIR does not fallback to Serial.print directly but to `serial_echopair_PGM` and there is no overload for `uint8_t` so which one of the other overload will be selected by the compiler isn't clear. When cast, it's clear which one will be used. If I write the type punning version, I'll also remove the cast, as it'll be clear in all cases.

rhapsodyv commented:
It would be good to at least test the compilation with all examples from the Configurations repository.

X-Ryl669 commented:
It's going to take a huge time to build all of them. I'm already running `make tests-all-local` and it takes more than an hour to finish on my computer (it's faster to let Github run them although it's not exactly the same).
So, until I consider the PR ready, I'll avoid this. When it's ready, I'll try to run as many as I can.

X-Ryl669 committed:
Fix the new reported errors due to bad encoding base of numbers

X-Ryl669 committed:
Replace SERIAL_PRINT with SERIAL_ECHO wherever appropriate

X-Ryl669 committed:
Restore varargs macros for ECHOPAIR since it gives the smallest value
I've failed to get a smaller binary size with variadic template. Whatever the mean, the resulting binary is always bigger than a basic macro unrolling/inlining.
So, I'm reverting the previous changes since it does not improve anything.

X-Ryl669 commented:
I'm not currently able to build all configuration from Marlin because I don't know what platformio environment to use. I've written a script to automate this, but it's missing additional information to work. If any of you has access to the Marlin's Configuration repository, feel free to add the missing `platfomio.env` file containing the expected environment to build. Once this is done, we'll be able to run `$ buildroot/ci-check/build_all_configs.sh bugfix-2.0.x` to build all configurations known to Marlin and check it does not break.

X-Ryl669 commented:
Concerning this PR, I think I'm done with it for now. I was not able to reduce binary footprint for SERIAL_ECHOPAIR to more than what it is actually with macro. I've tried with type punning (using a union) so that only one version of SERIAL_ECHOPAIR exists but it produced larger code than macros.  I've tried with type erasing too but again, it failed reducing the binary footprint. 

I think if we intended to gain less code size than now, we should instead rework the SERIAL_ECHOPAIR idea so something more or less like printf. It might be less efficient, but it'll take a lot less size. Yet, it's a too huge work for me to perform in a PR.

X-Ryl669 committed:
Fix DUE building with USB

qwewer0 commented:
Compiled it on `STM32F103RC_btt_512K`

Before:
```
RAM:   [===       ]  29.4% (used 14464 bytes from 49152 bytes)
Flash: [====      ]  45.0% (used 235804 bytes from 524288 bytes)
```

After:
```
RAM:   [===       ]  29.4% (used 14464 bytes from 49152 bytes)
Flash: [=====     ]  45.0% (used 235972 bytes from 524288 bytes)
```

X-Ryl669 commented:
@qwewer0 Can you post your configuration, please ?

qwewer0 commented:
@X-Ryl669 Sorry, should have known that it is best to include it.

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/5927451/Configuration.zip)

thinkyhead committed:
Misc updates

thinkyhead commented:
> I'm not currently able to build all configuration from Marlin because I don't know what platformio environment to use.

You can just use `mftest -a` and it will pick the correct environment.

thinkyhead committed:
Some README corrections

X-Ryl669 commented:
Is there a reason you're spelling nibble nybble ? I think the former is more frequent than the latter. 
I'll try the `mftest` trick tomorrow.

thisiskeithb commented:
> Is there a reason you're spelling nibble nybble ? I think the former is more frequent than the latter.

There's a mix of "nibble" and "nybble" spelling throughout Marlin.

thinkyhead committed:
Fix SERIAL_ECHOLIST for AVR

thinkyhead committed:
More tweaks

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20985

thinkyhead committed:
format comments

thinkyhead commented:
> There's a mix of "nibble" and "nybble" spelling throughout Marlin.

Some of our external contributions spell it "nibble." But there's a general preference here at the lab for "nybble."

thinkyhead commented:
If we can't use `char` inside any `SERIAL_ECHO` now, then we'll also need to fix lines like this:
```cpp
SERIAL_ECHOLNPAIR("Setting up encoder on ", axis_codes[encoderAxis], " axis, addr = ", address);
```

thinkyhead committed:
Assume char/uchar is numeric, unless...

thinkyhead commented:
Instead of splitting up lines like that to use `SERIAL_CHAR` I'm thinking of adding a new synthetic type like…
```cpp
typedef struct { char c; } serial_char_t;
#define AS_CHAR(C) serial_char_t({ C });
```
…and then the code can just be…
```cpp
SERIAL_ECHOLNPAIR("Setting up encoder on ", AS_CHAR(axis_codes[encoderAxis]), " axis, addr = ", address);
```

thinkyhead committed:
Use MSG macros where possible

thinkyhead committed:
Drop unnecessary int casting

thinkyhead committed:
clean up more casts

thinkyhead committed:
Fix char wrapper

thinkyhead committed:
single char

thinkyhead committed:
oops

thinkyhead committed:
Get uchar too…

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20985

thinkyhead committed:
Pass uchar sign to int

qwewer0 commented:
Increased a bit again. I'm not complaining, just giving information.

```
Bugfix-2.0.x:
RAM:   [===       ]  29.4% (used 14464 bytes from 49152 bytes)
Flash: [====      ]  45.0% (used 235804 bytes from 524288 bytes)

This:
RAM:   [===       ]  29.4% (used 14464 bytes from 49152 bytes)
Flash: [=====     ]  45.0% (used 236060 bytes from 524288 bytes)
```

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/5927451/Configuration.zip)

thinkyhead commented:
> Increased a bit again. I'm not complaining, just giving information.

Thanks! The size of the build is indeed one of the things we'll be evaluating as this gets refined.

thinkyhead commented:
Results look pretty reasonable here building with PlatformIO…
```
AVR sample build (old):

RAM:   [========  ]  82.4% (used 6750 bytes from 8192 bytes)
Flash: [========= ]  90.5% (used 229786 bytes from 253952 bytes)

AVR sample build (new):

RAM:   [========  ]  82.4% (used 6750 bytes from 8192 bytes)
Flash: [========= ]  90.4% (used 229694 bytes from 253952 bytes)

    -92 bytes (smaller)



LPC sample build (old):

RAM:   [===       ]  28.0% (used 9164 bytes from 32736 bytes)
Flash: [===       ]  29.3% (used 139036 bytes from 475136 bytes)

LPC sample build (new):

RAM:   [===       ]  28.0% (used 9164 bytes from 32736 bytes)
Flash: [===       ]  29.3% (used 139036 bytes from 475136 bytes)

     Same size



Ender 3 V2 (old):

RAM:   [==        ]  15.7% (used 10280 bytes from 65536 bytes)
Flash: [===       ]  30.6% (used 160292 bytes from 524288 bytes)

RAM:   [==        ]  15.7% (used 10280 bytes from 65536 bytes)
Flash: [===       ]  30.6% (used 160188 bytes from 524288 bytes)

    -104 bytes (smaller)
```

X-Ryl669 commented:
It's like a Heisinopt. Depending on where you live in the quantum universe, it reduces or increases your build :smile:

X-Ryl669 committed:
Missed a PGM output optimization

X-Ryl669 commented:
@qwewer0... 
So to sum up:
## Current bugfix builds for your configuration:
```
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  29.4% (used 14448 bytes from 49152 bytes)
Flash: [====      ]  44.3% (used 232204 bytes from 524288 bytes)
```

## This PR build with your configuration
```
Building .pio/build/STM32F103RC_btt_512K/firmware.bin
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  29.4% (used 14448 bytes from 49152 bytes)
Flash: [====      ]  44.3% (used 232460 bytes from 524288 bytes)
```
So it's bigger for you, but less than what you listed above. To understand where the increase is coming from, I've restored the macro code for the SERIAL_XXX macro only, so I can rule out the other changes that are required:

## Other change without the template replacing macro:
```
Building .pio/build/STM32F103RC_btt_512K/firmware.bin
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  29.4% (used 14448 bytes from 49152 bytes)
Flash: [====      ]  44.3% (used 232244 bytes from 524288 bytes)
```
So the impact of the fixes is ~40 bytes. The impact of the template code is 216 bytes. 
I'm continuing the investigations to explain where those bytes are spent...

X-Ryl669 commented:
Looking at the bugfix binary, the serial print functions consumes:
```
$ ~/.platformio/packages/toolchain-gccarmnoneeabi/bin/arm-none-eabi-objdump -tT -C .pio/build/STM32F103RC_btt_512K/firmware.elf | grep "::print" | grep Serial
/home/cyril/.platformio/packages/toolchain-gccarmnoneeabi/bin/arm-none-eabi-objdump: .pio/build/STM32F103RC_btt_512K/firmware.elf: not a dynamic object
0800c2ce  w    F .text	00000024 SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::printNumber(long, unsigned char)
0800c4a4  w    F .text	0000000a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(unsigned char, PrintBase)
0800c4c8  w    F .text	0000000a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(unsigned int, PrintBase)
0800c314  w    F .text	0000000a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(int, PrintBase)
0800c2f2  w    F .text	0000000a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(char, PrintBase)
0800c350  w    F .text	000000e0 SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::printFloat(double, unsigned char)
0800ce58  w    F .text	00000012 SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(EnsureDouble, int)
0800c278  w    F .text	00000056 SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::printNumber(unsigned long, unsigned char)

$ ~/.platformio/packages/toolchain-gccarmnoneeabi/bin/arm-none-eabi-objdump -tT -C .pio/build/STM32F103RC_btt_512K/firmware.elf | grep "::print" | grep Serial | awk '{ x = strtonum("0x" $5); printf("%d\n", x); } ' | paste -s -d+|bc
404
``` 
So 404 bytes.

Sidenote: What's strange is that it still contains `Print::print` which should never appear anymore in the binary...

The current PR gives:
```
~/.platformio/packages/toolchain-gccarmnoneeabi/bin/arm-none-eabi-objdump -tT -C .pio/build/STM32F103RC_btt_512K/firmware.elf | grep "::print" | grep Serial
08011f00 l     F .text	0000004a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::printNumber(unsigned long, unsigned char) [clone .part.1]
0800c480  w    F .text	00000024 SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::printNumber(long, unsigned char)
0800c438  w    F .text	0000000a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(unsigned char, PrintBase)
0800c45c  w    F .text	0000000a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(unsigned int, PrintBase)
0800c4c8  w    F .text	0000000a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(int, PrintBase)
0800c4a4  w    F .text	0000000a SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::print(char, PrintBase)
0800c2d0  w    F .text	000000e0 SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::printFloat(double, unsigned char)
0800c278  w    F .text	00000056 SerialBase<MultiSerial<BaseSerial<MarlinSerial>, ForwardSerial<USBSerial>, (unsigned char)0> >::printNumber(unsigned long, unsigned char)
```
It consumes 460 bytes. If you spot the difference, there is a copy of `printNumber(unsigned long, unsigned char)`. This should never happen due to ODR. 
Something's strange with the compiler or the compile options here. The code is not the problem I think.

X-Ryl669 commented:
Indeed, it's due to [partial inlining specialization](https://developers.redhat.com/blog/2014/10/29/rhel7-gcc-optimizations-partial-inlining/). I thought the function would be too huge for the compiler to inline. Notice that it's **not** the `signed long` version that's forwarding to the `unsigned long` version that's inlined but the huge `unsigned long` version.

X-Ryl669 commented:
And... 
Applying this patch:
```
diff --git a/Marlin/src/core/serial_base.h b/Marlin/src/core/serial_base.h
index 795cf772ac..81403b55f2 100644
--- a/Marlin/src/core/serial_base.h
+++ b/Marlin/src/core/serial_base.h
@@ -132,7 +132,7 @@ struct SerialBase {
   FORCE_INLINE void println(double c)              { println(c, 2); }

   // Print a number with the given base
-  void printNumber(unsigned long n, const uint8_t base) {
+  NO_INLINE void printNumber(unsigned long n, const uint8_t base) {
     if (!base) return; // Hopefully, this should raise visible bug immediately

     if (n) {
@@ -158,7 +158,7 @@ struct SerialBase {
   }

   // Print a decimal number
-  void printFloat(double number, uint8_t digits) {
+  NO_INLINE void printFloat(double number, uint8_t digits) {
     // Handle negative numbers
     if (number < 0.0) {
       write('-');
```
Gives:
```
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  29.4% (used 14448 bytes from 49152 bytes)
Flash: [====      ]  44.2% (used 231844 bytes from 524288 bytes)
```
for this PR.

X-Ryl669 commented:
It seems it also work for me:
```
RAM:   [=======   ]  74.6% (used 48880 bytes from 65536 bytes)
Flash: [====      ]  44.5% (used 233340 bytes from 524288 bytes)
```
40 more bytes to shave off :smile:

qwewer0 commented:
> ```
> diff --git a/Marlin/src/core/serial_base.h b/Marlin/src/core/serial_base.h
> index 795cf772ac..81403b55f2 100644
> --- a/Marlin/src/core/serial_base.h
> +++ b/Marlin/src/core/serial_base.h
> @@ -132,7 +132,7 @@ struct SerialBase {
>    FORCE_INLINE void println(double c)              { println(c, 2); }
> 
>    // Print a number with the given base
> -  void printNumber(unsigned long n, const uint8_t base) {
> +  NO_INLINE void printNumber(unsigned long n, const uint8_t base) {
>      if (!base) return; // Hopefully, this should raise visible bug immediately
> 
>      if (n) {
> @@ -158,7 +158,7 @@ struct SerialBase {
>    }
> 
>    // Print a decimal number
> -  void printFloat(double number, uint8_t digits) {
> +  NO_INLINE void printFloat(double number, uint8_t digits) {
>      // Handle negative numbers
>      if (number < 0.0) {
>        write('-');
> ```

```
bugfix:
RAM:   [===       ]  29.4% (used 14464 bytes from 49152 bytes)
Flash: [====      ]  45.0% (used 235804 bytes from 524288 bytes)

Patch:
RAM:   [===       ]  29.4% (used 14464 bytes from 49152 bytes)
Flash: [=====     ]  45.1% (used 236668 bytes from 524288 bytes)

With that two line changed:
RAM:   [===       ]  29.4% (used 14464 bytes from 49152 bytes)
Flash: [=====     ]  45.0% (used 236060 bytes from 524288 bytes)
```

X-Ryl669 commented:
There's something wrong with your build (or you haven't send me your exact configuration file). Please check you have updated packages (in particular toolchain). I've built your with your configuration above and I'm always getting smaller binaries than you. Maybe you need to clean your build for leftover ?
 
If you think you've done all of these, please post the result of:
```
~/.platformio/packages/toolchain-gccarmnoneeabi/bin/arm-none-eabi-objdump -tT -C .pio/build/STM32F103RC_btt_512K/firmware.elf | grep "::print" | grep Serial
```

X-Ryl669 commented:
See, there's something different in the RAM usage between you and me, and this can only result from different configuration. There is no change in RAM usage in this PR.

qwewer0 commented:
> There's something wrong with your build (or you haven't send me your exact configuration file). Please check you have updated packages (in particular toolchain). I've built your with your configuration above and I'm always getting smaller binaries than you. Maybe you need to clean your build for leftover ?
> 
> If you think you've done all of these, please post the result of:
> 
> ```
> ~/.platformio/packages/toolchain-gccarmnoneeabi/bin/arm-none-eabi-objdump -tT -C .pio/build/STM32F103RC_btt_512K/firmware.elf | grep "::print" | grep Serial
> ```

Don't know how to do that. I will try it if you can guide me. Maybe discord would be a better place for the chit-chat?

But here is the compilation (if this even helps at all):

```
Processing STM32F103RC_btt_512K (platform: ststm32@~10.0; board: genericSTM32F103RC; framework: arduino)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Tool Manager: toolchain-gccarmnoneeabi @ 1.70201.0 is already installed
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/genericSTM32F103RC.html
PLATFORM: ST STM32 (10.0.1) > STM32F103RC (48k RAM. 256k Flash)
HARDWARE: STM32F103RCT6 72MHz, 48KB RAM, 512KB Flash
DEBUG: Current (blackmagic) External (blackmagic, cmsis-dap, jlink, stlink)
PACKAGES:
 - framework-arduinoststm32-maple 3.10000.201129 (1.0.0)
 - tool-stm32duino 1.0.2
 - toolchain-gccarmnoneeabi 1.70201.0 (7.2.1)
Converting Marlin.ino
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 29 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <SoftwareSerialM> 1.0.0
|-- <TMCStepper> 0.7.1
|   |-- <SoftwareSerialM> 1.0.0
|-- <U8glib-HAL> 0.4.3
|   |-- <Wire> 1.0
|-- <STM32ADC> 1.0
|-- <EEPROM>
|-- <USBComposite for STM32F1> 0.99
|-- <Wire> 1.0
|-- <Servo(STM32F1)> 1.1.2
Building in release mode
...
Linking .pio\build\STM32F103RC_btt_512K\firmware.elf
Building .pio\build\STM32F103RC_btt_512K\firmware.bin
Checking size .pio\build\STM32F103RC_btt_512K\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  29.4% (used 14464 bytes from 49152 bytes)
Flash: [=====     ]  45.0% (used 236060 bytes from 524288 bytes)
================================================================================================= [SUCCESS] Took 302.92 seconds =================================================================================================
Environment           Status    Duration
--------------------  --------  ------------
STM32F103RC_btt_512K  SUCCESS   00:05:02.925
```

qwewer0 commented:
> See, there's something different in the RAM usage between you and me, and this can only result from different configuration. There is no change in RAM usage in this PR.

I always had that amount of ram used. I will clean install vscode and the other stuff later, hope it will help.

X-Ryl669 commented:
Seems like, except for `tool-stm32duino`, we've the same versions. I can't explain the differences.
I'm using the configuration file from [this post](https://github.com/MarlinFirmware/Marlin/pull/20985#issuecomment-773840038).

> Don't know how to do that. I will try it if you can guide me

If you are using VSCode, you can open a PIO terminal and paste the command line above, you can then post the result of the command line here. Else, if you are using linux, you can do the same in a terminal.
If you are using Windows, I guess you'll need to trigger a bash terminal somehow, since Microsoft Powershell will not understand the command above. In that case, I'm sorry I can't really help.


```
Processing STM32F103RC_btt_512K (platform: ststm32@~10.0; board: genericSTM32F103RC; framework: arduino)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/genericSTM32F103RC.html
PLATFORM: ST STM32 (10.0.1) > STM32F103RC (48k RAM. 256k Flash)
HARDWARE: STM32F103RCT6 72MHz, 48KB RAM, 512KB Flash
DEBUG: Current (blackmagic) External (blackmagic, cmsis-dap, jlink, stlink)
PACKAGES:
 - framework-arduinoststm32-maple 3.10000.201129 (1.0.0)
 - tool-stm32duino 1.0.1
 - toolchain-gccarmnoneeabi 1.70201.0 (7.2.1)
Converting Marlin.ino
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 29 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <SoftwareSerialM> 1.0.0
|-- <TMCStepper> 0.7.1
|   |-- <SoftwareSerialM> 1.0.0
|-- <U8glib-HAL> 0.4.3
|   |-- <Wire> 1.0
|-- <STM32ADC> 1.0
|-- <EEPROM>
|-- <USBComposite for STM32F1> 0.99
|-- <Wire> 1.0
|-- <Servo(STM32F1)> 1.1.2
Building in release mode
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/HAL.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/HAL_SPI.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/MarlinSerial.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/SPI.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/Servo.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/dogm/u8g_com_stm32duino_swspi.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/eeprom_bl24cxx.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/eeprom_flash.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/eeprom_if_iic.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/eeprom_sdcard.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/eeprom_wired.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/fast_pwm.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/msc_sd.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/onboard_sd.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/sdio.cpp.o
Compiling .pio/build/STM32F103RC_btt_512K/src/src/HAL/STM32F1/tft/tft_fsmc.cpp.o
Linking .pio/build/STM32F103RC_btt_512K/firmware.elf
Checking size .pio/build/STM32F103RC_btt_512K/firmware.elf
Building .pio/build/STM32F103RC_btt_512K/firmware.bin
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  29.4% (used 14448 bytes from 49152 bytes)
Flash: [====      ]  44.2% (used 231844 bytes from 524288 bytes)
================================================================================================================= [SUCCESS] Took 5.64 seconds =================================================================================================================

Environment           Status    Duration
--------------------  --------  ------------
STM32F103RC_btt_512K  SUCCESS   00:00:05.635
```

X-Ryl669 commented:
You don't need to clean VScode. You can clean the project via `platformio run -t clean`. I don't remember exactly, but if you have Platformio extension, I think there's a button in the status bar for that.

qwewer0 commented:
So, I'm not sure about this:
```
~/.platformio/packages/toolchain-gccarmnoneeabi/bin/arm-none-eabi-objdump -tT -C .pio/build/STM32F103RC_btt_512K/firmware.elf | grep "::print" | grep Serial
grep : The term 'grep' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:129
+ ... mp -tT -C .pio/build/STM32F103RC_btt_512K/firmware.elf | grep "::prin ...
+                                                              ~~~~
    + CategoryInfo          : ObjectNotFound: (grep:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

And after the cleanup, using my https://github.com/MarlinFirmware/Marlin/pull/20985#issuecomment-773840038 config.

```
Bugfix:
RAM:   [===       ]  29.4% (used 14448 bytes from 49152 bytes)
Flash: [====      ]  44.3% (used 232204 bytes from 524288 bytes)

(already committed) Patch:
RAM:   [===       ]  29.4% (used 14448 bytes from 49152 bytes)
Flash: [====      ]  44.3% (used 232460 bytes from 524288 bytes)

The two line changed:
RAM:   [===       ]  29.4% (used 14448 bytes from 49152 bytes)
Flash: [====      ]  44.2% (used 231844 bytes from 524288 bytes)
```

It seems that I used a bit different config, sorry for the inconvenience.

X-Ryl669 commented:
Ok, that was crazy!

For what it's worth, I've found why there's still a `Print::print` in your binary. It comes from `Marlin/src/lcd/dogm/lcdprint_u8g.cpp`, more specifically the `u8g.print()` lines in there (as stated by:
```
~/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: .pio/build/STM32F103RC_btt_512K/FrameworkArduino/IPAddress.cpp.o: reference to _ZN5Print5printEc
~/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: .pio/build/STM32F103RC_btt_512K/FrameworkArduino/Print.cpp.o: definition of _ZN5Print5printEc
~/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: .pio/build/STM32F103RC_btt_512K/src/src/lcd/dogm/lcdprint_u8g.cpp.o: reference to _ZN5Print5printEc
```
`u8g` happens to be `U8GLIB_ST7920_128X64_RRD` in your configuration and this is a SPI instance. SPI instance still use the Arduino's print stuff, so we end up with both print method here. 

I'm not sure if there are many platform using `print` with SPI's instance (it's a strange practice I think to output ASCII text on SPI), but for those, the saving in this PR will not cover the dual print code overhead. I can't fix for them. 

At least the mystery is solved!

X-Ryl669 committed:
Don't let the compiler inline on its own, it's doing wrongly.
Remove a leftover code from a bad merge in another PR

X-Ryl669 commented on Marlin/src/HAL/STM32F1/MarlinSerial.h:
```diff
@@ -49,28 +45,6 @@ struct MarlinSerial : public HardwareSerial {
       nvic_irq_set_priority(c_dev()->irq_num, UART_IRQ_PRIO);
     }
   #endif
-
```
This shouldn't have been included in the bugfix branch. It's useless now, let's remove it.

qwewer0 commented:
The +16 bytes of RAM was because I had `CANCEL_OBJECTS` enabled on my end.

rhapsodyv commented:
Could this be related with the serial refactoring?

#21010

thinkyhead commented:
Let's see what shakes out as a result of this general improvement.

thinkyhead commented:
Unfortunately this PR invented its own concept for `SERIAL_ECHOLIST` taking a string as its first argument and in the process eliminated `SERIAL_ECHOLIST_N`, which takes a count and a list of values. Note that `SERIAL_ECHOLIST_N` was only ever used in one file throughout all of Marlin. So now in order to do debugging in that file I am forced to spend extra time creating a new `SERIAL_ECHOLIST_N` passing an empty string, I guess … and extra time composing this message admonishing contributors to be _please strive to be less clever and make smaller incremental changes_ so that we can more easily evaluate them.

As long as we're here, I have asked participants in #21010 to report on the current situation, and if there are still problems we will have to do a deeper dive to understand how else this PR may have unintentionally altered behavior, timing, buffering, etc.
